pkg/gadgets/traceloop/tracer/tracer.go

```diff
@@ -0,0 +1,436 @@
+	ret       uint64   `json:"ret,omitempty"`
+	mountNsID uint64   `json:"mountnsid,omitempty"`
+	retval    int      `json:"ret,omitempty"`
+}
+
+type syscallEventContinued struct {
+	timestamp uint64 `json:"timestamp,omitempty"`
+	index     uint8  `json:"index,omitempty"`
+	param     string `json:"params,omitempty"`
+}
+
+func NewTracer(enricher gadgets.DataEnricher) (*Tracer, error) {
+	t := &Tracer{
+		enricher: enricher,
+	}
+
+	spec, err := loadTraceloop()
+	if err != nil {
+		return nil, fmt.Errorf("failed to load ebpf program: %w", err)
+	}
+
+	syscalls, err = gatherSyscalls()
+	if err != nil {
+		return nil, fmt.Errorf("error gathering syscall: %w", err)
+	}
+
+	// Fill the syscall map with specific syscall signatures.
+	syscallsMapSpec := spec.Maps["syscalls"]
+	for name, def := range syscallDefs {
+		nr, err := libseccomp.GetSyscallFromName(name)
+		if err != nil {
+			return nil, fmt.Errorf("cannot get syscall number of %q: %w", name, err)
+		}
+
+		syscallsMapSpec.Contents = append(syscallsMapSpec.Contents, ebpf.MapKV{
+			Key:   uint64(nr),
+			Value: unsafe.Pointer(&def[0]),
+		})
+	}
+
+	if err := spec.LoadAndAssign(&t.objs, nil); err != nil {
+		return nil, fmt.Errorf("failed to load ebpf program: %w", err)
+	}
+
+	t.enterLink, err = link.AttachRawTracepoint(link.RawTracepointOptions{
+		Name:    "sys_enter",
+		Program: t.objs.IgTraceloopE,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("error opening tracepoint: %w", err)
+	}
+
+	t.exitLink, err = link.AttachRawTracepoint(link.RawTracepointOptions{
+		Name:    "sys_exit",
+		Program: t.objs.IgTraceloopX,
+	})
+	if err != nil {
+		t.enterLink = gadgets.CloseLink(t.enterLink)
+
+		return nil, fmt.Errorf("error opening tracepoint: %w", err)
+	}
+
+	t.spec = spec
+
+	return t, nil
+}
+
+func (t *Tracer) Stop() {
+	t.enterLink = gadgets.CloseLink(t.enterLink)
+	t.exitLink = gadgets.CloseLink(t.exitLink)
+
+	t.readers.Range(func(key, _ any) bool {
+		mntnsID, ok := key.(uint64)
+		if ok {
+			t.Detach(mntnsID)
+		}
+
+		return ok
+	})
+
+	t.objsLock.Lock()
+	t.objs.Close()
+	t.objsLock.Unlock()
+}
+
+func (t *Tracer) Attach(mntnsID uint64) error {
+	outerMapSpec := t.spec.Maps["map_of_perf_buffers"]
+	innerMapSpec := outerMapSpec.InnerMap
+
+	innerBufferSpec := innerMapSpec.Copy()
+	innerBufferSpec.Name = fmt.Sprintf("perf_buffer_%d", mntnsID)
+
+	// 1. Create inner Map as perf buffer.
+	innerBuffer, err := ebpf.NewMap(innerBufferSpec)
+	if err != nil {
+		return fmt.Errorf("error creating inner map: %w", err)
+	}
+
+	// 2. Use this inner Map to create the perf reader.
+	perfReader, err := perf.NewReaderWithOptions(innerBuffer, gadgets.PerfBufferPages*os.Getpagesize(), perf.ReaderOptions{WriteBackward: true, OverWritable: true})
+	if err != nil {
+		return fmt.Errorf("error creating perf ring buffer: %w", err)
+	}
+
+	// 3. Add the inner map's file descriptor to outer map.
+	t.objsLock.Lock()
+	err = t.objs.MapOfPerfBuffers.Put(mntnsID, innerBuffer)
+	t.objsLock.Unlock()
```

Comment:
I don't think we need a mutex to update a ebpf.Map in parallel.

Reply:
I think we can completely remove `t.objsLock`: it seems to protect `t.objs.MapOfPerfBuffers` from `.Put` and `.Delete` in parallel but that is just calling the bpf() syscall which the kernel should be able to run in parallel.